contracts/contracts/coordination/FlatRateFeeModel.sol

```diff
@@ -3,34 +3,113 @@
 pragma solidity ^0.8.0;
 
 import "./IFeeModel.sol";
+import "./Coordinator.sol";
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
 
 /**
  * @title FlatRateFeeModel
  * @notice FlatRateFeeModel
  */
-contract FlatRateFeeModel is IFeeModel {
+contract FlatRateFeeModel is IFeeModel, Ownable {
+    using SafeERC20 for IERC20;
+
     IERC20 public immutable currency;
     uint256 public immutable feeRatePerSecond;
+    Coordinator public immutable coordinator;
+
+    uint256 public totalPendingFees;
+    mapping(uint256 => uint256) public pendingFees;
 
-    constructor(IERC20 _currency, uint256 _feeRatePerSecond) {
+    constructor(
+        Coordinator _coordinator,
+        IERC20 _currency,
+        uint256 _feeRatePerSecond
+    ) Ownable(msg.sender) {
         require(_feeRatePerSecond > 0, "Invalid fee rate");
         currency = _currency;
         feeRatePerSecond = _feeRatePerSecond;
+        coordinator = _coordinator;
     }
 
     function getRitualInitiationCost(
-        address[] calldata providers,
+        uint256 numberOfProviders,
         uint32 duration
     ) public view returns (uint256) {
-        uint256 size = providers.length;
         require(duration > 0, "Invalid ritual duration");
-        require(size > 0, "Invalid ritual size");
-        return feeRatePerSecond * size * duration;
+        require(numberOfProviders > 0, "Invalid ritual size");
+        return feeRatePerSecond * numberOfProviders * duration;
     }
 
     // TODO: Validate if this is enough to remove griefing attacks
     function feeDeduction(uint256, uint256) public pure returns (uint256) {
         return 0;
     }
+
+    function processRitualPayment(
+        address initiator,
+        uint32 ritualId,
+        uint256 numberOfProviders,
+        uint32 duration
+    ) external override {
+        require(msg.sender == address(coordinator), "Only coordinator can call process payment");
+        uint256 ritualCost = getRitualInitiationCost(numberOfProviders, duration);
+        require(ritualCost > 0, "Invalid ritual cost");
+        totalPendingFees += ritualCost;
```

Comment:
does this balance cover encryptor credits and the base fee?

Reply:
this is initial fee model, not yet subscription. so no credits or base fee, just flat rate